Scripts re-added for PR & code review.

hepcat72 opened the pull request:
Files checked in: src/splitseq_combine.sh src/splitseq_mkref.sh src/splitseq_run.sh

hepcat72 committed:
Scripts re-added for PR & code review.

Files checked in: src/splitseq_combine.sh src/splitseq_mkref.sh src/splitseq_run.sh

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
```
Consider adding `set -euxo pipefail` for a more robust script. This might break something, but shouldn't be too hard to fix. It's kinda like Perl's strict mode. (See https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/ for some explanation).

hepcat72 replied:
Should I add those options to line 17 where `set -eu` currently is? I put it there to permit the mkdir of the outer directory to complain that it already exists.  Alternatively, I could put that mkdir inside an if statement.

hepcat72 replied:
I just edited the existing line.  Let me know after I push if you'd prefer I add a conditional so it can be at the top.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+fastas=$2        #space-delimited, e.g. "hg38.fa mm10.fa"
+gtfs=$3          #space-delimited, e.g. "hg38.gtf mm10.gtf"
+
+splitseq_venv="/Genomics/grid/users/rleach/local/splitseq"
```
Consider creating an `environment.yml` file that creates a conda env with all required deps. This would be a bit more portable.

hepcat72 replied:
I tried `conda env export > splitseq.yml`, but it has `prefix: /Genomics/grid/users/rleach/local/miniconda2/envs/splitseq` in the resulting file.  I'll see if I can figure out how to have that omitted tomorrow.

hepcat72 replied:
For documentation pruposes, this was informative: https://github.com/conda/conda/issues/4339

lparsons replied:
I almost always create the `environment.yml` by hand.

lparsons replied:
Also, I didn't see the conda env file in the PR, did you mean to omit it? Would be useful I think.

hepcat72 replied:
I haven't pushed the changes yet. I have implemented them locally.  Just testing to confirm the scripts still work with the changes.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+
+mkdir "${refdir}"
+
+echo ./bin/python ./bin/split-seq mkref \
```
Using `set -x` will output commands that are executed to `stdout`.

hepcat72 replied:
Cool.  I'll add that.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+    --output_dir "${refdir}" \
+    --nthreads 16
+
+./bin/python ./bin/split-seq mkref \
```
`./bin/python` isn't necessary if the virtual environment is activated first. In general, I'd rather leave it to the user to setup the environment (see above).

hepcat72 replied:
Well, the user does set up the environment using virtualenv, as described in the nplcadmindocs split seq analysis page.  That strategy didn't put python in the beginning of my path, which is why I did it this way.  I had included that in this repo so that they could be reviewed together so that this would be clear, but I think what you're suggesting is better.  Will do.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+splitseq_refs="${splitseq_venv}/splitseq_refs"
+refdir="${splitseq_refs}/"`echo "${genome_names}" | perl -pne 's/ /_/g;'`
+
+cd "${splitseq_venv}"
```
Activate environments instead of `cd` to them. Also, generally, I'd prefer the user to setup the environment and have the script assume things are in the path. That way, the user can use modules, conda env, or python venv as they see fit.

hepcat72 replied:
Will do.  I'm still a noob with this environment stuff.  Will do.

hepcat72 replied:
So there's no conda package for split-seq.  If I don't do a cd, what's the best way to supply the split-seq script to their python call?

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+splitseq_venv="/Genomics/grid/users/rleach/local/splitseq"
+splitseq_runs="${splitseq_venv}/splitseq_runs"
+
+cd "${splitseq_venv}"
```
Same general comments re env as mkref.

hepcat72 replied:
Got it.

lparsons reviewed (changes requested):
Looks like a nice set of scripts. Good work sorting out everything we need to run the pipeline and also setting up some nice default names, etc.

I made a few specific comments on mkref that apply to all. Generally, I'd have three recommendations.

1. Provide instructions on setting up the environment, but don't do it in each script. Let the user decide how they want to provide the dependencies. As a convenience, an `environment.yml` file that uses conda to setup all dependencies would be great.

2. Create a `README.md` file that includes basic instructions, excluding things specific to our environment.

3. Get rid of the docs directory, it will be in NPLC admin docs. I'm fine with us just editing that as we see fit.

hepcat72 committed:
Removed docs and added a .gitignore.

Files checked in: doc/nplcadmindocs.txt

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
```
Add `set -euxo pipefail`

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+splitseq_venv="/Genomics/grid/users/rleach/local/splitseq"
+splitseq_runs="${splitseq_venv}/splitseq_runs"
+
+cd "${splitseq_venv}"
```
I would remove the references to the venv. It would be nice to keep that separate from the scripts. If the user provides the dependencies via venv, great, but they shouldn't be required to, imo anyway.

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+
+mkdir "${splitseq_runs}"
+
+set -eu
```
Remove `set -eu` (put `set -euxo pipefail` at the top.

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+
+mkdir "${splitseq_runs}/${run_id}"
+
+echo ./bin/python ./bin/split-seq combine \
```
Don't specify paths to python or split-seq. Let the user's environment define that. Unless of course, split-seq doesn't end up in the path after a `pip install` I'm just not sure how that works, perhaps I need to start trying this out. Is there a simple toy example I can run somewhere on `gen-comp2`? Perhaps in `Genomics/biocomp`? Might be helpful so I can make more cogent and relevant comments. :wink:

lparsons replied:
The `echo` line isn't required if [`set -x`](https://vaneyckt.io/posts/safer_bash_scripts_with_set_euxo_pipefail/#set-x) is used, since all commands are printed before executing.

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
```
Change to `#!/usr/bin/bash` or `#!/usr/bin/env bash`.

hepcat72 replied:
Will do.

lparsons commented on src/splitseq_combine.sh:
```diff
@@ -0,0 +1,49 @@
+
+module load STAR/2.7.3a
+module load samtools/1.10
+module load python/3.6.4
```
I suggest that the module stuff be left out or perhaps moved to a `setup_module_env.sh` script to allow flexibility. For example, I would prefer to create a conda env for all of these dependencies and not use `module` or `venv` at all.

hepcat72 replied:
Good idea. I'll include in conda.

hepcat72 replied:
I removed the modules and included them in the conda env.  Tests worked and I pushed.  Oh wait, I have yet to update the yaml, but this item is essentially done..

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Change to `#!/usr/bin/bash` or `#!/usr/bin/env bash`.

hepcat72 replied:
Will do.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+
+module load STAR/2.7.3a
+module load samtools/1.10
+module load python/3.6.4
```
Remove `venv` and `module` references.

hepcat72 replied:
Got it.

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
+
```
Add `set -euxo pipefail`

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+
+mkdir "${splitseq_refs}"
+
+set -eu
```
Remove `set -eu`

lparsons commented on src/splitseq_mkref.sh:
```diff
@@ -0,0 +1,44 @@
+
+mkdir "${refdir}"
+
+echo ./bin/python ./bin/split-seq mkref \
```
Avoid explicit paths to executables.
Remove `echo` statement.

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
```
Change to `#!/usr/bin/bash` or `#!/usr/bin/env bash`.
Add `set -euxo pipefail`

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+splitseq_venv="/Genomics/grid/users/rleach/local/splitseq"
+splitseq_runs="${splitseq_venv}/splitseq_runs"
+
+cd "${splitseq_venv}"
```
Remove `venv` reference.

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+
+mkdir "${splitseq_runs}"
+
+set -eu
```
Remove `set -eu`

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+
+module load STAR/2.7.3a
+module load samtools/1.10
+module load python/3.6.4
```
Remove/move module refs

hepcat72 replied:
Where to?

hepcat72 replied:
Found the other comment answering this Q.

lparsons commented on src/splitseq_run.sh:
```diff
@@ -0,0 +1,60 @@
+    --chemistry "${chemistry}" \
+    --genome_dir "${ssrefdir}" \
+    --nthreads 16 \
+    ${sample_args}
```
Avoid explicit paths to python, etc.
Remove `echo`.

lparsons reviewed (changes requested):
Looking good. I made basically the same set of comments on each script (sorry if that's a lot of noise).

I mentioned this in one of the comments, but I'll put it here for clarity. Perhaps I need to start trying this out. Is there a simple toy example I can run somewhere on `gen-comp2`? Perhaps in `Genomics/biocomp`? Might be helpful so I can make more cogent and relevant comments. :wink:

hepcat72 committed:
Implemented changes suggested in PR.

Files checked in: src/splitseq_combine.sh src/splitseq_mkref.sh src/splitseq_run.sh

hepcat72 committed:
Added star and samtools to the environment.

Files checked in: environment.yml

hepcat72 commented:
OK, this is ready for review round 2.  If you'd prefer a bare bones environment yaml file, let me know.  I think I like having everything in there that I know works.  If some update breaks something, people won't be affected, though that admittedly does create more overhead for updating the repo...

hepcat72 commented:
I also updated the nplcadmindocs page.

hepcat72 commented:
What am I missing here? This shows "1 change requested", but I don't see any unresolved issues.  I need to do a run for Wei, so I'm just going to run what I've got in this branch for him.

lparsons reviewed (approved):
I don't think you're missing anything. Just didn't notice you were ready. Sorry, I have to sort out GitHub notifications for repos not in an org. They go to my Gmail account, and subsequently get overlooked.
